util/microservice-sdk/js/lib/kafka/Producer.js

```diff
@@ -0,0 +1,166 @@
+    this.isReady = false;
+    this.producer = new Kafka.Producer(config);
+
+    // This will set a callback to be executed everytime a message is published.
+    const resolveOnDeliveryReportBind = Producer.resolveOnDeliveryReport.bind(this);
+    this.producer.on('delivery-report', resolveOnDeliveryReportBind);
+    this.producer.setPollInterval(100);
+    logger.debug('... a Kafka producer was successfully created.', TAG);
+  }
+
+  /**
+   * Connects the Producer to a Kafka cluster.
+   *
+   * This function will wait for 5 seconds in order to successfully connect to
+   * a broker instance. If there is any problem, then it will time out or
+   * node-rdkakfa will emit a event.error event, both of them triggering a
+   * rejection of the returned promise.
+   *
+   * @returns { Promise } A promise which will be resolved if this Producer is
+   * successfully connected to a Kafka broker, or reject if any errors occur.
+   */
+  connect() {
+    logger.info('Connecting the producer...', TAG);
+    const readyPromise = new Promise((resolve, reject) => {
+      const timeoutTrigger = setTimeout(() => {
+        logger.error('Failed to connect the producer.', TAG);
+        reject(new Error('timed out'));
+      }, 5000);
+
+      this.producer.on('ready', () => {
+        logger.debug('Producer is ready.', TAG);
+        clearTimeout(timeoutTrigger);
+        this.isReady = true;
+        return resolve();
+      });
+
+      this.producer.on('event.error', (error) => {
+        logger.debug(`Error while creating producer: ${error}`, TAG);
+        return reject(error);
+      });
+    });
+
+    this.producer.connect(null);
+    return readyPromise;
+  }
+
+  /**
+   * Produce a message to a Kafka broker.
+   *
+   * This function will create a message to be published to a Kafka broker. If
+   * message key is set, then all messages with the same key will use the same
+   * partition. If the message partition is set, then it will be used when
+   * producing the message to a Kafka broker.
+   *
+   *
+   * @param { string } topic The topic to which the message will be published
+   * @param { string } message The actual message
+   * @param { string } key (optional) -The key to be used to selected a partition.
+   * @param { number } partition (optional) - The particion to be used when publishing the
+   * message.
+   */
+  produce(topic, message, key = null, partition = null) {
+    return new Promise((resolve, reject) => {
+      const timeStamp = Date.now();
+
+      const buffer = Buffer.from(message);
+
+      const callback = { resolve, reject };
+
+      // The callback object passed here will be used as an opaque object, that
+      // is, the delivery report function (set in the constructor) will be
+      // invoked with this object - it will resolve or reject the function
+      // depending on the fallout of the produce operation.
+      this.producer.produce(topic, partition, buffer, key, timeStamp, { callback });
```

Comment:
I'm not sure what happens if the dr_cb is not true. Won't the promise be resolved nor rejected?